### PR TITLE
Remove 'tensor' of a joined TensorMap if possible

### PR DESCRIPTION
**Pull request by PicoCentauri**
Adds the functionality to `join` to remove the `"tensor"` column name while joining TensorMaps.

Due to torch-script I can not use a try-catch block like

```python
try:
    return remove_labels_column(tensor_joined, name="tensor", axis=axis)
except EquistoreError:
    return tensor_joined
```

Instead, I have to check the uniqueness on the Python level before calling `remove_labels_column`. If someone has an idea to get around the his I am happy to change this.

### CODE REVIEW

File: python/equistore-operations/equistore/operations/join.py [hunk 1]
```diff
@@ -126,7 +160,22 @@ def join(tensors: List[TensorMap], axis: str):
...
+            new_names = np.delete(block.samples, -1, axis=1)
+        elif axis == "properties":
+            n_values = len(block.properties)
+            # "tensor" name is inserted at the begininng of the Labels dimensions
+            new_names = np.delete(block.properties, 0, axis=1)
+
+        n_values_unique = len(np.unique(new_names, axis=0))
+
+        if n_values != n_values_unique:
+            return tensor_joined
```
Comment (jwa7) [inline python/equistore-operations/equistore/operations/join.py:179]: Perhaps `unique_metadata` can be used here?

```suggestion
    if len(equistore.unique_metadata(tensor_joined, axis=axis, name="tensor")) > 1:
        return tensor_joined
```

Comment (jwa7) [reply python/equistore-operations/equistore/operations/join.py:179]: Hmm apologies maybe not: see https://github.com/lab-cosmo/equistore/pull/314#discussion_r1273274348

File: python/equistore-operations/equistore/operations/join.py [hunk 2]
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
...
+        labels_1 = ["n"], [0, 2, 3]
+        labels_2 = ["n"], [1, 4, 5]
+
+       will lead to::
+
+        joined_labels = ["n"],
+                        [0, 2, 3, 1, 4, 5]
+
+    2. If Labels names are the same but the values are not unique, a new dimension
+       ``"tensor"`` is added to the names::
```
Comment (agoscinski) [inline python/equistore-operations/equistore/operations/join.py:36]:
```suggestion
       ``"tensor"`` is added at the beginning to the names which determines from which TensorMap the label entry originally comes from ::
```

maybe a comment on where the new name is added, always at the beginning?

File: python/equistore-operations/equistore/operations/join.py [hunk 3]
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
...
-    ``join`` will create an additional label `tensor` specifiying the original index in
-    the list of `tensor_maps`.  If `sample`/`property` names are not the same in all
-    `tensor_maps` they will be unified with a general name ``"property"``.
+    Possible clashes of the meta data like samples/properties names are resolved by one
+    of the three following strategies:
+
+    1. If Labels names are the same and the values are unique, we keep the names and
+       join the values::
+
+        labels_1 = ["n"], [0, 2, 3]
```
Comment (agoscinski) [inline python/equistore-operations/equistore/operations/join.py:27]: This could be actually super nice doctests. not enforcing this, but would be nice

File: python/equistore-operations/equistore/operations/join.py [hunk 4]
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
...
     samples dimension.
 
-    ``join`` will create an additional label `tensor` specifiying the original index in
-    the list of `tensor_maps`.  If `sample`/`property` names are not the same in all
-    `tensor_maps` they will be unified with a general name ``"property"``.
+    Possible clashes of the meta data like samples/properties names are resolved by one
+    of the three following strategies:
+
+    1. If Labels names are the same and the values are unique, we keep the names and
+       join the values::
```
Comment (agoscinski) [inline python/equistore-operations/equistore/operations/join.py:25]: why two double colons at the end? is this related to sphinx?

File: python/equistore-operations/equistore/operations/join.py [hunk 5]
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
...
+
+        labels_1 = ["n"], [0, 2, 3]
+        labels_2 = ["n"], [0, 2]
+
+       will lead to::
+
+            joined_labels = ["tensor", "n"],
+                            [[0, 0], [0, 2], [0, 3], [1, 0], [1, 2]]
+
+    3. If Labels names are different we change the names to ("tensor", "property"). This
```
Comment (agoscinski) [inline python/equistore-operations/equistore/operations/join.py:46]:
```suggestion
    3. If Labels names are different we change the names to ("tensor", "property"). The column "property" corresponds to the ith label entry TensorMap specified by tensor. This
```

File: python/equistore-operations/equistore/operations/join.py [hunk 6]
```diff
@@ -126,7 +160,22 @@ def join(tensors: List[TensorMap], axis: str):
     else:
         tensor_joined = tensor.keys_to_properties("tensor")
 
-    # TODO: once we have functions to manipulate meta data we can try to
-    # remove the `tensor` label entry after joining.
-
-    return tensor_joined
+    # Remove "tensor" Labels dimension if a new Labels of all blocks would be unique
+    # Due to toch script we can use a try-except block but have to check in advance if
```
Comment (agoscinski) [inline python/equistore-operations/equistore/operations/join.py:164]:
```suggestion
    # Due to toch script we cant use a try-except block but have to check in advance if
```

**Review by agoscinski — COMMENTED** [review]
If we input a TensorMap with a label column name "tensor" or "property" can we raise an error already in this function? Something like 
```
raise ValueError(f"join cannot be applied on TensorMaps with a column name 'tensor' in the TensorMap keys or in targeted axis {axis}. Please rename the corresponding column first using ....")
```

File: python/equistore-operations/equistore/operations/join.py [hunk 7]
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
...
+    3. If Labels names are different we change the names to ("tensor", "property"). This
+       case is only supposed to happen when joining in the property dimension, hence the
+       choice of names::
+
+            labels_1 = ["a"], [0, 2, 3]
+            labels_2 = ["b", "c"], [[0, 0], [1, 2]]
+
+       will lead to::
+
+            joined_labels = ["tensor", "property"],
```
Comment (agoscinski) [inline python/equistore-operations/equistore/operations/join.py:55]: What do you think about using a notation for "private" variables in python for metadata column names that can exclusively only used equistore for operations "_eqs_tensor" and "_eqs_property". To be honest, I don't have any other ideas of applications than this function, but it could be already here useful.

File: python/equistore-operations/equistore/operations/join.py [hunk 8]
```diff
@@ -17,9 +18,42 @@ def join(tensors: List[TensorMap], axis: str):
     `properties` dimension and for ``axis='samples'`` they will be the along the
```
Comment (agoscinski) [inline python/equistore-operations/equistore/operations/join.py:18]:
```
`axis='properties'`` it will be the `tensor_maps` will be joined along the
```
->
```
`axis='properties'`` the tensor maps in `tensors` will be joined along the
```

File: python/equistore-operations/equistore/operations/join.py (hunk 4 above)
Comment (PicoCentauri) [reply python/equistore-operations/equistore/operations/join.py:25]: Yes this renders the next paragraph as code.

File: python/equistore-operations/equistore/operations/join.py (hunk 3 above)
Comment (PicoCentauri) [reply python/equistore-operations/equistore/operations/join.py:27]: Yes, I though the same. I think I can do it!

File: python/equistore-operations/equistore/operations/join.py (hunk 7 above)
Comment (PicoCentauri) [reply python/equistore-operations/equistore/operations/join.py:55]: Hmm I think private names will make everything harder to follow.

File: python/equistore-operations/equistore/operations/join.py (hunk 1 above)
Comment (PicoCentauri) [reply python/equistore-operations/equistore/operations/join.py:179]: Yes as we talked yesterday I will try to use Labels::intersection.

File: python/equistore-operations/equistore/operations/join.py [hunk 9]
```diff
@@ -7,6 +7,7 @@
 from equistore.core import Labels, TensorBlock, TensorMap
 
 from ._utils import _check_same_keys
+from .manipulate_dimension import remove_dimension
 
 
 def join(tensors: List[TensorMap], axis: str):
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply python/equistore-operations/equistore/operations/join.py:13]: Yes it is more `concatenate` but I would say we keep it like this and can rename the bunch when we have more operations.

**Comment by PicoCentauri** [comment]
I added an new boolean parameter `remove_tensor_name` to control if the tensor dimension should be removed or not. We can also leave this out and completly and only put a comment that the user can remove the tensor using the `equistore.remove_dimension`. To be honest I am more and more in favor to do this. What do the reviewers think @jwa7, @agoscinski  and @Luthaf 

Also I added your the examples you want @agoscinski. Hope they are fine and understandable. I tried to keep the code to create the tensors as small as possible.

**Review by jwa7 — COMMENTED** [review]
I like the having the bool option to remove, but think we should have a default to False. Thoughts?

File: python/equistore-operations/equistore/operations/join.py (hunk 7 above)
Comment (agoscinski) [reply python/equistore-operations/equistore/operations/join.py]: They actually do this in pandas for their merge function
https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.merge.html
> **indicator: bool or str, default False**
>
> If True, **adds a column to the output DataFrame called “_merge” with information on the source of each row**. The column can be given a different name by providing a string argument. The column will have a Categorical type with the value of “left_only” for observations whose merge key only appears in the left DataFrame, “right_only” for observations whose merge key only appears in the right DataFrame, and “both” if the observation’s merge key is found in both DataFrames.

**Comment by agoscinski** [comment]
> I added an new boolean parameter remove_tensor_name to control if the tensor dimension should be removed or not. We can also leave this out and completly and only put a comment that the user can remove the tensor using the equistore.remove_dimension. To be honest I am more and more in favor to do this. What do the reviewers think @jwa7, @agoscinski and @Luthaf

Given the discussion in https://github.com/lab-cosmo/equistore/issues/315 we will break the API of join anyway, I dont think it matters much what we choose now. Let's just go with an option and set it to False as @jwa7 suggests it.

**Review by agoscinski — APPROVED** [review]
